presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,537 @@
+                    outputPositionCount++;
+                }
+                else {
+                    outputPositionCount -= filter.getPrecedingPositionsToFail();
+                    i += filter.getSucceedingPositionsToFail();
+                }
+            }
+        }
+        else if (nullsAllowed) {
+            outputPositionCount = positionCount;
+        }
+        else {
+            outputPositionCount = 0;
+        }
+
+        allNulls = true;
+        return positions[positionCount - 1] + 1;
+    }
+
+    private void skip(int items)
+            throws IOException
+    {
+        if (dataStream == null) {
+            presentStream.skip(items);
+        }
+        else if (presentStream != null) {
+            int lengthToSkip = presentStream.countBitsSet(items);
+            dataStream.skip(lengthStream.sum(lengthToSkip));
+        }
+        else {
+            dataStream.skip(lengthStream.sum(items));
+        }
+    }
+
+    /**
+     * skip the dataStream from the streamPosition to the current position
+     * increments the lengthIndex and
+     * calculate the length sum to the skipped.
+     */
+    private void skipData(int start, int items)
+            throws IOException
+    {
+        int dataToSkip = 0;
+        for (int i = 0; i < items; i++) {
+            if (isNullVector == null || !isNullVector[start + i]) {
+                dataToSkip += lengthVector[lengthIndex];
+                lengthIndex++;
+            }
+        }
+        dataStream.skip(dataToSkip);
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+
+        if (positionCount == outputPositionCount) {
+            Block block = new VariableWidthBlock(positionCount, slice, offsetVector, Optional.ofNullable(includeNulls ? nulls : null));
+            nulls = null;
+            offsetVector = null;
+            data = null;
+            slice = null;
+            return block;
+        }
+
+        boolean[] nullsCopy = null;
+        if (includeNulls) {
+            nullsCopy = new boolean[positionCount];
+        }
+
+        compactValues(positions, positionCount, includeNulls, nullsCopy);
+
+        return new VariableWidthBlock(positionCount, slice, offsetVector, Optional.ofNullable(includeNulls ? nullsCopy : null));
```

Comment:
you need to set data, slice and offsets to null - otherwise, you'll be sharing them with the block

Reply:
is this not the same for other readers as well. will follow up on this tomorrow.